### PR TITLE
Time phases

### LINKED ISSUE #933: Add something like rustc's -Z time-passes
This would give a high level overview of how long each of our phases are taking, and help inform us where we should generally be focusing our performance investigations on various test cases.

### PR DESCRIPTION
**Pull request by jhod0**
New module `time_phases` with an RAII timer which prints to stderr, and command-line flag `--time-phases` to enable it.

Fixes #933

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -63,7 +63,9 @@ mod clang;
 mod features;
 mod ir;
 mod parse;
+//mod raii_timer;
```
Comment (fitzgen) [inline src/lib.rs]: Nitpick: remove commented out code.

File: src/lib.rs [hunk 2]
```diff
@@ -1403,8 +1416,15 @@ impl<'ctx> Bindings<'ctx> {
             options.clang_args.push(f.name.to_str().unwrap().to_owned())
         }
 
+        let time_phases = options.time_phases;
         let mut context = BindgenContext::new(options);
-        try!(parse(&mut context));
+        {
+            // TODO this is a dummy value for now, should be command line flag
+            //let time_passes = false;
```
Comment (fitzgen) [inline src/lib.rs]: Nitpick: these comments can both be deleted.

File: src/time_phase.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
+use std::io::{self, Write};
+use std::time::{Instant, Duration};
+
+
+/// RAII timer to measure how long phases take.
+// TODO does it need output parameter? or always print to stdout
```
Comment (fitzgen) [inline src/time_phase.rs]: Always printing to stderr is fine.

File: src/time_phase.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    output: bool,
+    name: &'a str,
+    start: Instant,
+}
+
+
+impl<'a> Timer<'a> {
+    /// Creates a Timer with the given name, and starts it.
+    ///
+    /// By default, the output stream is None, and nothing is printed
```
Comment (fitzgen) [inline src/time_phase.rs]: I think this comment is out of date, possible referring to a time when the output was configurable.

File: src/lib.rs [hunk 5]
```diff
@@ -774,6 +776,13 @@ impl Builder {
         self
     }
 
+    /// Set whether or not to time bindgen phases, and print
+    /// information to stderr.
+    pub fn time_phases(mut self) -> Self {
```
Comment (fitzgen) [inline src/lib.rs]: For consistency with our other on/off features and options, this should have a `doit: bool` parameter, and set `self.options.time_phases = doit;`

File: src/lib.rs [hunk 6]
```diff
@@ -63,7 +63,9 @@ mod clang;
 mod features;
 mod ir;
 mod parse;
+//mod raii_timer;
 mod regex_set;
+mod time_phase;
```
Comment (fitzgen) [inline src/lib.rs]: Let's rename the module to `timer`.

File: src/ir/context.rs [hunk 7]
```diff
@@ -731,6 +745,7 @@ impl<'ctx> BindgenContext<'ctx> {
     /// Iterate over all items and replace any item that has been named in a
     /// `replaces="SomeType"` annotation with the replacement type.
     fn process_replacements(&mut self) {
+        let _t = self.timer("process_replacements");
```
Comment (fitzgen) [inline src/ir/context.rs]: Yep, I see you figured out how to do RAII variables that are only used for their `Drop` without getting unused variable warnings :)

File: src/codegen/mod.rs [hunk 8]
```diff
@@ -3478,6 +3478,7 @@ impl CodeGenerator for ObjCInterface {
 }
 
 pub fn codegen(context: &mut BindgenContext) -> Vec<P<ast::Item>> {
+    let _t = context.timer("codegen");
```
Comment (fitzgen) [inline src/codegen/mod.rs]: This is going to make codegen's timer include all of our analyses that get run from `gen`, which we don't really want here.

Let's move the timer into the closure passed to `gen` below.

File: src/ir/context.rs (hunk 7 above)
Comment (jhod0) [reply src/ir/context.rs]: It's a learning process :)

**Comment by jhod0** [comment]
@fitzgen I think all fixed

**Comment by fitzgen** [comment]
@bors-servo r+

Perfect! Thanks @jhod0 !

**Pull request merged**